File: 189_rotate_array/rotate.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        n = len(nums)

        if n == k:
            return

        if n == 1:
            return
        r = n - (k % n)
        temp = nums[::]

        for l in range(n):
            if r < n:
                nums[l] = temp[r]
                nums[r] = temp[l]
                if k > n and r == n:
                    r = n - k
            else:
                if abs(l - k) <= n:
                    nums[l] = temp[l - k]
            r += 1

File: 189_rotate_array/rotate.ts
/**
 Do not return anything, modify nums in-place instead.
 */
function rotate(nums: number[], k: number): void {
  const n = nums.length;
  k = k % n;

  if (n === 1 || k === 0) {
    return;
  }

  const temp = [...nums];

  for (let i = 0; i < k + 1; i++) {
    nums[i] = temp[i + n - k];
  }

  for (let i = k; i < n; i++) {
    nums[i] = temp[i - k];
  }
}
